### PR TITLE
[Conditions] Request, Subscribe to, and Provide Telemetry

### PR DESCRIPTION
**Pull request by davetsay**
## Overview
Addresses #2709, #2736
- Requests LAD from telemetry endpoints and calculates output
- Maintains one subscription to each telemetry endpoint
  - Criteria only listen to their endpoints
- Provides telemetry

## Author Checklist
| | |
| --- | :---: |
| Changes address original issue? | Y |
| Unit tests included and/or updated with changes? | Y |
| Command line build passes? | Y |
| Changes have been smoke-tested? | Y |

### CODE REVIEW

File: src/plugins/condition/ConditionManager.js [hunk 1]
```diff
@@ -31,13 +31,47 @@ export default class ConditionManager extends EventEmitter {
...
+            console.log('no subscription to remove');
+            return;
+        }
+
+        this.subscriptions[id]();
+        delete this.subscriptions[id];
+    }
+
+    load() {
+        return this.loaded;
```
Comment (deeptailor) [inline src/plugins/condition/ConditionManager.js]: Does this return a boolean? The name of the function implies that it loads something

File: src/plugins/condition/criterion/TelemetryCriterion.js [hunk 2]
```diff
@@ -116,34 +120,35 @@ export default class TelemetryCriterion extends EventEmitter {
...
-    /**
-     *  Calls an unsubscribe function returned by subscribe() and deletes any initialized data
-     */
-    unsubscribe() {
-        //unsubscribe from telemetry source
-        if (typeof this.subscription === 'function') {
-            this.subscription();
-        }
-        delete this.subscription;
+        return this.objectAPI.get(this.objectAPI.makeKeyString(this.telemetry))
```
Comment (deeptailor) [inline src/plugins/condition/criterion/TelemetryCriterion.js]: you can pass the identifier directly into the objectAPI.get, no need to turn it into a keystring

File: src/plugins/condition/criterion/TelemetryCriterion.js [hunk 3]
```diff
@@ -116,34 +120,35 @@ export default class TelemetryCriterion extends EventEmitter {
...
+                        id: this.id,
+                        data: this.formatData(latestDatum)
+                    };
+                });
+            });
     }
 
     destroy() {
-        this.unsubscribe();
+        this.on(`subscription:${this.telemetryObjectIdAsString}`, this.handleSubscription);
```
Comment (deeptailor) [inline src/plugins/condition/criterion/TelemetryCriterion.js]: would this be this.off?

File: src/plugins/condition/ConditionManager.js (hunk 1 above)
Comment (davetsay) [reply src/plugins/condition/ConditionManager.js]: i'll see if i can clarify

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 2 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: done. thanks.

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 3 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: yes it would. thanks.

File: src/plugins/condition/ConditionManager.js (hunk 1 above)
Comment (davetsay) [reply src/plugins/condition/ConditionManager.js]: fixed.

File: src/plugins/condition/Condition.js [hunk 4]
```diff
@@ -172,7 +186,7 @@ export default class ConditionClass extends EventEmitter {
         let found = this.findCriterion(criterion.id);
         if (found) {
             this.criteria[found.index] = criterion.data;
-            this.subscribe();
+            // this.subscribe();
```
Comment (shefalijoshi) [inline src/plugins/condition/Condition.js]: Remove this comment if it's not being used anymore.

File: src/plugins/condition/ConditionManager.js [hunk 5]
```diff
@@ -31,13 +31,41 @@ export default class ConditionManager extends EventEmitter {
...
+        this.composition.on('remove', this.removeTelemetry, this);
+        this.subscriptions = {};
         this.initialize();
 
         this.stopObservingForChanges = this.openmct.objects.observe(this.conditionSetDomainObject, '*', (newDomainObject) => {
             this.conditionSetDomainObject = newDomainObject;
         });
     }
 
+    addTelemetry(endpoint) {
```
Comment (shefalijoshi) [inline src/plugins/condition/ConditionManager.js]: Probably should rename this from `addTelemetry` to `subscribeToTelemetry` and for `removeTelemetry` to `unsubscribeFromTelemetry`

File: src/plugins/condition/ConditionManager.js [hunk 6]
```diff
@@ -31,13 +31,41 @@ export default class ConditionManager extends EventEmitter {
         this.conditionSetDomainObject = conditionSetDomainObject;
         this.timeAPI = this.openmct.time;
         this.latestTimestamp = {};
+        this.composition = this.openmct.composition.get(conditionSetDomainObject);
+        this.composition.on('add', this.addTelemetry, this);
```
Comment (shefalijoshi) [inline src/plugins/condition/ConditionManager.js]: Looks like you're subscribing to any telemetry that's added to the conditionSet. But we should only be subscribing to telemetry that's being used by conditions.

File: src/plugins/condition/criterion/TelemetryCriterion.js [hunk 7]
```diff
@@ -116,34 +119,35 @@ export default class TelemetryCriterion extends EventEmitter {
...
-    /**
-     *  Calls an unsubscribe function returned by subscribe() and deletes any initialized data
-     */
-    unsubscribe() {
-        //unsubscribe from telemetry source
-        if (typeof this.subscription === 'function') {
-            this.subscription();
-        }
-        delete this.subscription;
+        return this.objectAPI.get(this.telemetryObjectIdAsString)
```
Comment (shefalijoshi) [inline src/plugins/condition/criterion/TelemetryCriterion.js]: We already have the telemetry object as `this.telemetryObject`. Can you reuse that instead of making a new call get the telemetry object?

File: src/plugins/condition/criterion/TelemetryCriterion.js [hunk 8]
```diff
@@ -44,33 +44,36 @@ export default class TelemetryCriterion extends EventEmitter {
         this.operation = telemetryDomainObjectDefinition.operation;
         this.input = telemetryDomainObjectDefinition.input;
         this.metadata = telemetryDomainObjectDefinition.metadata;
-        this.subscription = null;
-        this.telemetryObjectIdAsString = null;
-        this.objectAPI.get(this.objectAPI.makeKeyString(this.telemetry)).then((obj) => this.initialize(obj));
+        this.telemetryObjectIdAsString = this.objectAPI.makeKeyString(this.telemetry);
+        this.on(`subscription:${this.telemetryObjectIdAsString}`, this.handleSubscription);
+        this.objectAPI.get(this.telemetryObjectIdAsString).then((obj) => this.initialize(obj));
```
Comment (shefalijoshi) [inline src/plugins/condition/criterion/TelemetryCriterion.js]: This change to use `this.telemetryObjectIdAsString` instead of `this.telemetry` will fail hard when you add a criterion as the telemetry is not yet set.
Either revert this change or ensure that a criterion class isn't created unless the telemetry has been chosen for it.

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 7 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: Sure. This was done with the intent of removing this.telemetryObject. If that is not needed we can reuse here.

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 8 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: Makes sense. (working backwards in replying to comments...) This was done because this.telemetryObject was going to be removed as not necessary for providing telemetry. If the telemetry object is still needed I can revert this.

File: src/plugins/condition/ConditionManager.js (hunk 5 above)
Comment (davetsay) [reply src/plugins/condition/ConditionManager.js]: sure

File: src/plugins/condition/ConditionManager.js (hunk 6 above)
Comment (davetsay) [reply src/plugins/condition/ConditionManager.js]: Correct. This was done intentionally to reduce calculations per discussion with @deeptailor. There shouldn't be too many instances of telemetry being added and not used. If this needs rework then we should open a new issue and discuss as a team.

File: src/plugins/condition/Condition.js (hunk 4 above)
Comment (davetsay) [reply src/plugins/condition/Condition.js]: removed

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 7 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: Actually. For now at least I'll leave this one as is. Even if this is being called twice I need to wait for this async call before requesting LAD. If I use this.telemetryObject as is it will not be loaded in time.

File: src/plugins/condition/criterion/TelemetryCriterion.js (hunk 8 above)
Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: Reverting to have telemetry object and setting in initialize function.

Comment (davetsay) [reply src/plugins/condition/criterion/TelemetryCriterion.js]: Reverted to use telemetryObject and initialize from that
